Add extra checks and logging to scribe; fix many warnings

wjt opened the pull request:
I saw a transient GPG verification failure when reformatting from a nightly ISO in a virtual machine. There was no useful information in the journal (beyond GPG's "BAD signature" output); re-running the process (without a reboot or anything) worked fine; and I manually GPG-verified the ISO, the squashfs within it, and the uncompressed image within that. Mysterious!

There isn't really much to go on. Maybe this was a transient read error from the USB disk on which the ISO is stored. Maybe a cosmic ray flipped a bit. But in case it happens again, here is some extra debug output and an extra coherence check in the scribe.

I also included many warning fixes, found by editing this project in GNOME Builder building against a non-Endless system.

wjt committed:
eos-installer.doap: update

wjt committed:
tests: add missing #include for memfd_create()

glibc >= 2.27 define this syscall. 025d89e updated our libglnx snapshot
to a version which includes this header to check for the existing
definition before defining a new version; but didn't update the function
which uses memfd_create() to include this header.

(The header exists in glibc < 2.27, so it's harmless to include it
there.)

wjt committed:
finished: add missing #include <locale.h>

This file uses setlocale() but did not include its header, instead
relying on it being included by some other header.

wjt committed:
Adapt to GCC 8's -Wcast-function-type

This new diagnostic produces warnings like:

  gis-scribe.c:1075:25: error: cast between incompatible function types
    from ‘gboolean (*)(GObject *, void *)’ {aka ‘int (*)(struct _GObject *, void *)’}
    to ‘gboolean (*)(void *)’ {aka ‘int (*)(void *)’} [-Werror=cast-function-type]

g_task_attach_source() takes a GSourceFunc which it passes to
g_source_set_callback(). However, the actual function type depends on
the specific GSource's implementation; the function pointer will be cast
back to the correct type later. So in this case the warning is (a)
spurious; and (b) forced on us by the GLib API. The workaround described
in the GCC documentation is to cast via (void (*) (void)); we employ
that here.

https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wcast-function-type

For the cast to GtkTickCallback, the warning is legit: the function did
have the wrong type. It was leaning on my belief that on all platforms
we care about, the ABI allows calling a function with extra arguments.
But it's easy to declare the function with the right type, and remove
the cast.

wjt committed:
install: adapt to GLib 2.56's type-preserving g_object_ref

In recent GLib, g_object_ref() has a macro wrapper that propagates its
argument type to its return type:

  /* Make reference APIs type safe with macros */
  #define g_object_ref(Obj)      ((__typeof__(Obj)) (g_object_ref) (Obj))

This lead to a warning where we assign a GObject * to a GFile *. Add an
explicit type check.

wjt committed:
main: avoid NULL-pointer dereference

Recent GCC warns about this.

(In fact, this function is an unused vestige of gnome-initial-setup. The
language will never be changed, so pages will only be "rebuilt" once, so
there will never be a current page, so destroy_pages_after() will never
be called. But this one-line fix will do until we have time to unpick
this unused code.)

wjt committed:
scribe: remove unused variable

wjt committed:
scribe: check expected number of bytes are teed

I saw a transient GPG verification failure when reformatting from a
nightly ISO in a virtual machine. There was no useful information in the
journal (beyond GPG's "BAD signature" output); re-running the process
(without a reboot or anything) worked fine; and I manually GPG-verified
the ISO, the squashfs within it, and the uncompressed image within that.
Mysterious!

In the absence of any clues, here's one coherence check we could add. In
theory this is redundant with the similar check on the total
uncompressed size in the write thread -- particularly on an ISO, where
we're verifying and writing the loop-mounted uncompressed image so the
two sizes are equal. But maintaining two additional counters shouldn't
hurt too much.

wjt committed:
scribe: log GPG command line

The only other reason for the transient GPG verification failure I can
think of is that for some reason we were passing the wrong signature
to GPG.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1072,7 +1072,7 @@ gis_scribe_begin_verify (GisScribe *self,
                                            cancellable);
   g_task_attach_source (task, task_data->stdout_source,
-                        (GSourceFunc) gis_scribe_gpg_progress);
+                        (GSourceFunc) (void (*)(void)) gis_scribe_gpg_progress);
```
I think upstream would probably be interested in a `G_SOURCE_FUNC()` casting macro for this. But to avoid dependency faff, you should go forth with `(void (*)(void))` here for now.

pwithnall replied:
*hopeful smile*

wjt replied:
MVP at https://gitlab.gnome.org/GNOME/glib/merge_requests/82

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1110,6 +1110,8 @@ gis_scribe_tee_data_free (GisScribeTeeData *data)
   g_slice_free (GisScribeTeeData, data);
 }
 
+G_STATIC_ASSERT (G_MAXSSIZE <= G_MAXUINT64);
```
What does this relate to? It’s not immediately obvious (from the diff view in GitHub) where you’re sticking a `gsize` value in a `guint64`.

⇒ Maybe add a comment

wjt replied:
It's in the function below but I think this check is (a) unnecessary (our images are certainly < 2 ** 62 - 1) or (b) incomplete (the gsize is repeatedly added to the guint64 so if I were really over-egging it I should add an overflow check). OK.

pwithnall replied:
Yeah, if you’re repeatedly adding the `gsize` to the `guint64` then it needs a runtime overflow check, if you care about that kind of thing.

wjt committed:
Update libglnx submodule URL

wjt commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1147,9 +1149,18 @@ gis_scribe_tee_thread (GTask            *task,
           break;
         }
+
+      bytes_teed += r;
```
I realised that 2 out of the 3 error paths above are not reached in the test suite, so I'll take a look at fixing that…

wjt replied:
The read and write-to-GPG paths are now tested (and worked all along).

wjt committed:
Test errors when reading from source image

wjt committed:
scribe: ensure first subtask error is reported

I had a report of an installation failing with:

> error writing image to self: Error writing to file descriptor: Broken
> pipe

which comes from the tee thread. I believe what actually happened is the
writer thread failed because the target hard disk is failing (as
indicated by a pop-up notification and the SMART information in GNOME
Disks), and we just picked the wrong error to display. Unfortunately the
message about subtask errors we're discarding was g_debug() and so would
not have appeared in the journal, even if I had been able to get the
journal from the reporter.

Looking at the code, on write error gis_scribe_write_thread() will call
g_task_return_error(), then close() the read end of the pipe coming from
the decompressor/tee thread, set its task as failed, and return. The tee
thread, on error, does something similar.

But when a task is running in a thread, g_task_return_error() doesn't
actually schedule the callback until the thread function returns. I
think this means that if the write thread happens to yield between
calling close() on the pipe from the tee thread, it's possible that the
tee task will signal its error on the main thread first, and we'd report
a less useful error.

To fix this, make each subtask explicitly set the error when it fails.
We also log all errors reported by subtasks when the main thread hears
about them.

All access to self->error was already guarded by the glorious mutex
because the write thread needs to wait for all other threads to complete
with success.

wjt committed:
Test what happens if GPG exits prematurely

One could quibble over what the error should be, but at least *some*
error should be reported, the first MiB zeroed, and so on.

wjt commented on tests/test-scribe.c:
```diff
@@ -787,6 +790,33 @@ main (int argc, char *argv[])
+      .image_path = image_path,
+      .signature_path = image_sig_path,
+      .gpg_path = "false",
+      .error_domain = G_IO_ERROR,
```
Oh right! This is racy! If the tee task notices its pipe-to-GPG is closed before the subprocess task notices "GPG" has exited, we'll get BROKEN_PIPE. But if (as happened) on Jenkins the race goes the other way, we'll get GIS_IMAGE_ERROR_VERIFICATION_FAILED. I think I don't care which error we get, just that *some* error is reported.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -228,6 +230,11 @@ gis_scribe_set_property (GObject      *object,
 
+    case PROP_GPG_PATH:
+      g_free (self->gpg_path);
+      self->gpg_path = g_value_dup_string (value);
```
Nitpick: Maybe add some assertions that the path is non-empty and absolute? I normally add such assertions in the `constructed` vfunc, since it’s guaranteed to be called after all construct-only properties are set, regardless of whether a value is actually provided for the property (or the default value is used).

wjt replied:
There's a non-NULL assertion in `constructed`. I'll add a check for absolution, and make it hold true.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -372,9 +375,14 @@ test_error (Fixture       *fixture,
+  if (fixture->data->error_domain != 0)
+    g_assert_error (error,
+                    fixture->data->error_domain,
+                    fixture->data->error_code);
```
Maybe add
```
else
  g_assert_nonnull (error);
```
?

wjt replied:
Oh yes! I meant to do this. Thanks.

pwithnall reviewed (commented):
A couple of small comments.

pwithnall reviewed (approved):
LGTM!
